### PR TITLE
Optimize MagGeoBuilder

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- Avoid extra work done for provenance tracking when looking up magnetic field related files.

The IB performance measurements for the NANO step shows that initializing the MagField object takes the majority of the startup time for the job.

#### PR validation:

Code compiles and performance measurements show an improvement.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
Before the change we see
```
            3.6  .........       1.45 / 1.62         magneticfield::MagGeoBuilder::build(cms::DDDetector const*) [261]
[296]       3.6       1.45       0.00 / 1.45       magneticfield::MagGeoBuilder::buildInterpolator(magneticfield::volumeHandle const*, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, MagneticFieldProvider<float>*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, MagneticFieldProvider<float>*> > >&)
            2.2  .........       0.88 / 0.91         edm::FileInPath::FileInPath(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) [429]
            1.4  .........       0.55 / 0.55         MFGridFactory::build(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, GloballyPositioned<float> const&) [646]
```

after 
```
            3.3  .........       1.30 / 1.46         magneticfield::MagGeoBuilder::build(cms::DDDetector const*) [273]
[324]       3.3       1.30       0.02 / 1.28       magneticfield::MagGeoBuilder::buildInterpolator(magneticfield::volumeHandle const*, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, MagneticFieldProvider<float>*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, MagneticFieldProvider<float>*> > >&)
            1.9  .........       0.76 / 0.76         MFGridFactory::build(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, GloballyPositioned<float> const&) [478]
            1.3  .........       0.49 / 0.49         edm::FileInPath::findFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) [680]
```

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
The underlying performance problem is `MagGeoBuilder` reads 9178 small files. E.g.

/cvmfs/cms-ib.cern.ch/nweek-02732/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_4_X_2022-05-10-1100/external/slc7_amd64_gcc10/data/MagneticField/Interpolation/data/grid_160812_3_8t/s10/grid.1049.bin

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
The failures in `test-das-selected-lumis` and `test_edmPickEvents` are along
```
DAS error: b'2022/05/13 01:26:19 ERROR failed to parse X509 proxy: crypto/tls: failed to parse key: asn1: syntax error: data truncated\n'
```
Maybe some glitch?

File: FWCore/Utilities/src/FileInPath.cc [hunk 1]
```diff
@@ -436,4 +440,22 @@ namespace edm {
 
   void FileInPath::disableFileLookup() { s_fileLookupDisabled = true; }
 
+  std::string FileInPath::findFile(const std::string& iFileName) {
+    // Find the file, based on the value of path variable.
+    auto pathElements = tokenize(searchPath(), ":");
+    for (auto const& element : pathElements) {
+      // Set the boost::fs path to the current element of
```
Comment (makortel) [inline FWCore/Utilities/src/FileInPath.cc:447]: Maybe time to update the comment in `initialize_()`.

**Comment by Dr15Jones** [comment]
Please test

**Comment by Dr15Jones** [comment]
ping

the unit test failure is from the IB.

**Comment by makortel** [comment]
+core

**Comment by jpata** [comment]
Was there any impact on memory usage?

**Comment by Dr15Jones** [comment]
> Was there any impact on memory usage?

This is essentially just a refactoring of the code without any changes to any memory structures. As such there is no reason for the memory footprint to have changed.

**Comment by jpata** [comment]
+reconstruction
- technical, speeds up loading the magnetic fields

**Comment by jpata** [comment]
> This is essentially just a refactoring of the code without any changes to any memory structures. As such there is no reason for the memory footprint to have changed.

Got it. On a quick read, I had the impression you are caching the magnetic fields, but I missed the point.

**Comment by qliphy** [comment]
+1

**Comment by qliphy** [comment]
merge

**Pull request merged**